cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/HierarchyTreeNode.tsx

```diff
@@ -0,0 +1,366 @@
+        childrenOfSelected = handleChildren(childrenOfSelected, item.children, uniqueId, 0);
+      }
+      const childToBeAdded = {
+        id: uniqueId,
+        path: item.path,
+        parentId: node.id,
+        name: item.name,
+        children: [],
+        type: item.type,
+      };
+      setShowMultiSelect(false);
+      setChildVisiblity(true);
+      newRecords.push(childToBeAdded);
+      if (childrenOfSelected) {
+        childrenOfSelected.forEach((child) => {
+          newRecords.push(child);
+        });
+      }
+    });
+    setJsonData(toJsonHandler(newRecords));
+    setRecords(newRecords);
+    setMultiSelectOptions([]);
+  };
+
+  const nameChangeHandler = (e) => {
+    const newRecords = [...records];
+    newRecords.forEach((record) => {
+      if (record.id === node.id) {
+        record.name = e.target.value;
+      }
+    });
+    setJsonData(toJsonHandler(newRecords));
+    setRecords(newRecords);
+  };
+
+  const handleAddNewRecordToParent = () => {
+    const newRecords = [...records];
+    newRecords.push({
+      id: uuidV4(),
+      parentId: node.id,
+      type: 'record',
+      children: [],
+      name: '',
+    });
+    setJsonData(toJsonHandler(newRecords));
+    setRecords(newRecords);
+    setChildVisiblity(true);
+  };
+
+  const recur = (...values) => ({ recur, values });
+
+  const loop = (f) => {
+    let a = f();
+    while (a && a.recur === recur) {
+      a = f(...a.values);
+    }
+    return a;
+  };
+
+  const removeFamily = (id = 0, nodes = []) =>
+    loop(([node, ...more] = nodes, s = new Set([id]), r = []) =>
+      node === undefined
+        ? r
+        : s.has(node.id) || s.has(node.parentId)
+        ? recur([...r, ...more], s.add(node.id), [])
+        : recur(more, s, [...r, node])
+    );
+
+  const handleDeleteElement = () => {
+    const newRecords = [...records];
+    const elements = removeFamily(node.id, newRecords);
+    setJsonData(toJsonHandler(elements));
+    setRecords(elements);
+  };
+
+  const ArrowIcon = (
+    <ButtonIcon
+      onClick={() => {
+        setChildVisiblity((value) => !value);
+        setShowMultiSelect(false);
+      }}
+    >
+      {childVisible ? <KeyboardArrowDownIcon /> : <KeyboardArrowRightIcon />}
+    </ButtonIcon>
+  );
+
+  const renderChildren = (option, selected) => {
+    return (
+      <div style={{ marginLeft: `${option.margin}px` }}>
+        <Checkbox checked={selected} />
+        {option.name}
+        {option.children &&
+          option.children.length !== 0 &&
+          option.children.map((child) => renderChildren(child, selected))}
+      </div>
+    );
+  };
+
+  return (
+    <React.Fragment>
+      {node.parentId !== null && (
+        <div className={classes.relative}>
+          <div className={classes.hierarchyIcon} />
+          <div className={classes.innerMostSiblingConnector} />
+        </div>
+      )}
+      <RowInputContainer data-cy="row">
+        <InputFieldWrapper>
+          {node.type === 'enum' ||
```

Comment:
Can we extract this into a separate functional component under hierarchy widget, that accepts the type and renders arrow based on the type.